text/trace/0000-sampling-propagation.md

```diff
@@ -0,0 +1,126 @@
+of a child span created in this context, use `2^log-count`.  A
+log-count of `0` corresponds with `(2^0)=1`, thus 0 conveys a context
+with probability 1.
+
+The sampling probability of a context is independent from whether it
+is sampled.  We consider it [useful to convey sampling probability
+even when unsampled]() as it can be used to estimate the potential
+overhead of starting new sampled traces.
+
+### Examples
+
+These are extended [from the W3C
+examples](https://www.w3.org/TR/trace-context/#examples-of-http-traceparent-headers):
+
+```
+Value = 01-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01-05
+base16(version) = 01
+base16(trace-id) = 4bf92f3577b34da6a3ce929d0e0e4736
+base16(parent-id) = 00f067aa0ba902b7
+base16(trace-flags) = 01  // sampled
+base16(log-count) = 05  // head probability is 2^-5.
+```
+
+```
+Value = 01-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-00-00
+base16(version) = 01
+base16(trace-id) = 4bf92f3577b34da6a3ce929d0e0e4736
+base16(parent-id) = 00f067aa0ba902b7
+base16(trace-flags) = 00  // not sampled
+base16(log-count) = 11 // head probability is 2^-17
+```
+
+We are able to express sampling probabilities as small as 2^-255 using
+just 3 bytes per `traceparent`.
+
+## Internal details
+
+The reasoning behind restricting the set of sampling rates is that it:
+
+- Lowers the cost of propagating head sampling probability
+- Makes math involving partial traces tractable.
+
+A use known as "inflationary sampling" from Google's Dapper system is
+documented in [OTEP 148](TODO).  This is is used to justify
+propagating the head sampling probability even when unsampled.
+
+[An algorithm for making statistical inferance from partially-sampled
+traces has been published](https://arxiv.org/pdf/2107.07703.pdf) that
+explains how to work with power-of-2 sampling rates.
+
+## Trade-offs and mitigations
+
+Restricting head sampling rates to powers of two does not limit tail
+Samplers from using arbitrary probabilities.
+
+Restricting head sampling rates to powers of two does not limit
+Samplers from using arbitrary effective probabilities over a period of
+time.  For example, choosing 1/2 sampling half of the time and 1/4
+sampling half of the time leads to an effective sampling rate of 3/8.
+
+## Prior art and alternatives
+
+Google's Dapper system propagated a field in its trace context called
+"inverse_probability", which is equivalent to adjusted count.  This
+proposal uses the base-2 logarithm of adjusted count to save space
+
+## Open questions
+
+This OTEP suggests how to modify the W3C trace context to accomodate
+sampling in OpenTelemetry.  [OTEP 148](TODO) suggests semantic
+conventions for encoding adjusted count in a Span, but neither text
+specifies how to modify the built-in Samplers to produce the proposed
+new `traceparent` field so that the `ParentBased` Sampler can
+correctly set the proposed `sampler.adjusted_count` attribute.  This
+will be future work.
```

Comment:
I support this proposal. A related question to this is whether we also want to extend `traceparent` by a geometrically distributed random value with a success probability of 1/2 as described in https://arxiv.org/pdf/2107.07703.pdf (section 2.8) and as needed for trace ID ratio based sampling (if we want to avoid hashing the trace ID for every span). This random value could also be represented by a single byte and would have the following meaning:

* 0 would indicate that only spans with sampling rate equal to 1 (with `log-count` = 0) are recorded. 
* 1 would indicate that only spans with a sampling rate greater than or equal to 1/2 (with `log-count` <= 1) are recorded. 
* 2 would indicate that only spans with a sampling rate greater than or equal to 1/4 (with `log-count` <= 2) are recorded. 
* 3 would indicate that only spans with a sampling rate greater than or equal to 1/8 (with `log-count` <= 3) are recorded. 
* ...

As a consequence, the sampled flag would become obsolete as it can be directly derived from a comparison of `log-count` with the random value. Since the geometrically distributed random value would require adding another field and hence also a further upgrade of the `traceparent` version, it should be discussed, if we would like to achieve both at once.

Btw, the proposed geometrically distributed random value is a discrete version of the `sampling score` as proposed in https://github.com/open-telemetry/oteps/pull/135.

Reply:
@oertl Thank you. 

> the sampled flag would become obsolete as it can be directly derived from a comparison of `log-count` with the random value

Your proposal seems to require that `TraceIDRatio` be the only probability sampler, which leads me to wonder why we would ever propagate the `is-sampled` bit. I think we have to keep a way for the parent to signal to the child that a span should be recorded without giving a probability, although these spans are difficult to count.

Your proposal suggests to me the use of a new flag to signal probabilistic sampling as opposed to non-probabilistic sampling. If we accept your proposal and keep the is-sampled bit for non-probability sampling, do we still need to propagate the head probability?

I will have to think about this a bit more. 

> discrete version of the `sampling score` as proposed in #135

This makes sense. It makes me wonder, though, if we have a way to generate the correct randomness, what's really keeping us from using random bits in TraceID? If we have random bits in the TraceID, the number-of-leading-zeros test can be applied directly, right?

Reply:
@oertl You proposed something related

> whether we also want to extend traceparent by a geometrically distributed random value

and I am trying to determine whether to include more text in this OTEP. It's tempting to call these independent solutions to independent problems, but maybe that is not the case.

One way of framing the problem is this. We want to ensure that the `sampler.adjusted_count` attribute proposed in #170 (previously #148) is always correct regardless of sampler configuration, as long as each participant acts in good faith. In this framing, the two problems are connected because we have to anticipate a system that uses a mixture of `AlwaysOn`, `AlwaysOff`, `Parent`, and `TraceIDRatio` samplers. 

At a minimum we need two new specifications involving W3C trace context and/or how it is used in OTel:

1. We have to propagate the head sampling probability when `Parent` is used
2. We have to specify a how to obtain randomness when `TraceIDRatio` is used

We also need to show that there is a reasonable expectation that `sampler.adjusted_count` is always set correctly if Samplers are updated according to this specification, no matter which one of them is used. I will work on a revision to this OTEP to clarify this intended outcome.

Reply:
@oertl I've been considering my own question above, 

> if we have a way to generate the correct randomness, what's really keeping us from using random bits in TraceID?

and I gather that the answer is -- it requires less randomness with your proposal. In the version where we add a single byte to convey randomness, the expected number of random bits required is 2 in this case (for the reader, I believe we're talking about the mean of a https://en.wikipedia.org/wiki/Geometric_distribution with p=0.5). Can you confirm?

What you propose, in practice, might take a buffer of random bits and for each new trace context root created it would set a new `random` field in the W3C trace context, as in:

```
traceparent: (version)-(trace_id)-(span_id)-(flags)-(log_count)(random)
```

with a new flag `randomized`, value 0x4 to support a new field.

```
// Logic for TraceID root sampler
 r := 0
 for {
     if nextRandomBit() {
       break // The expected value of r is 2
     }
    r++
}
// Set random to r for child contexts that will use their own TraceID probability
// Set is-sampled based on r and the root's configured TraceID probability
// Set log_count to the negative base-2 logarithm of the root's configured TraceID probability
```

Please confirm my understanding, thanks!

Reply:
(confirmed below: https://github.com/open-telemetry/oteps/pull/168#issuecomment-888369265)